# 02 — Data Understanding

Load the raw dataset, inspect structure, dtypes, summary statistics, and target distribution. Uses `src/data_validation.py` for the loading logic so the same function is reused in later automated checks.

In [1]:
import sys
sys.path.insert(0, '..')
import pandas as pd
import numpy as np

from src.utils import RAW_DATA_PATH, TARGET_COL, ID_COLS
from src.data_validation import load_raw_data

pd.set_option('display.max_columns', 40)
df = load_raw_data(RAW_DATA_PATH)
print(f"Shape: {df.shape}")
df.head()

[09:11:11] INFO - src.data_validation - Loaded raw data: 2149 rows x 35 columns


Shape: (2149, 35)


,PatientID,Age,Gender,Ethnicity,EducationLevel,BMI,Smoking,AlcoholConsumption,PhysicalActivity,DietQuality,SleepQuality,FamilyHistoryAlzheimers,CardiovascularDisease,Diabetes,Depression,HeadInjury,Hypertension,SystolicBP,DiastolicBP,CholesterolTotal,CholesterolLDL,CholesterolHDL,CholesterolTriglycerides,MMSE,FunctionalAssessment,MemoryComplaints,BehavioralProblems,ADL,Confusion,Disorientation,PersonalityChanges,DifficultyCompletingTasks,Forgetfulness,Diagnosis,DoctorInCharge
0,4751,73,0,0,2,22.927749,0,13.297218,6.327112,1.347214,9.025679,0,0,1,1,0,0,142,72,242.366840,56.150897,33.682563,162.189143,21.463532,6.518877,0,0,1.725883,0,0,0,1,0,0,XXXConfid
1,4752,89,0,0,0,26.827681,0,4.542524,7.619885,0.518767,7.151293,0,0,0,0,0,0,115,64,231.162595,193.407996,79.028477,294.630909,20.613267,7.118696,0,0,2.592424,0,0,0,0,1,0,XXXConfid
2,4753,73,0,3,1,17.795882,0,19.555085,7.844988,1.826335,9.673574,1,0,0,0,0,0,99,116,284.181858,153.322762,69.772292,83.638324,7.356249,5.895077,0,0,7.119548,0,1,0,1,0,0,XXXConfid
3,4754,74,1,0,1,33.800817,1,12.209266,8.428001,7.435604,8.392554,0,0,0,0,0,0,118,115,159.582240,65.366637,68.457491,277.577358,13.991127,8.965106,0,1,6.481226,0,0,0,0,0,0,XXXConfid
4,4755,89,0,0,0,20.716974,0,18.454356,6.310461,0.795498,5.597238,0,0,0,0,0,0,94,117,237.602184,92.869700,56.874305,291.198780,13.517609,6.045039,0,0,0.014691,0,0,1,1,0,0,XXXConfid


## 2.1 Structure & Data Types

In [2]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2149 entries, 0 to 2148
Data columns (total 35 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   PatientID                  2149 non-null   int64  
 1   Age                        2149 non-null   int64  
 2   Gender                     2149 non-null   int64  
 3   Ethnicity                  2149 non-null   int64  
 4   EducationLevel             2149 non-null   int64  
 5   BMI                        2149 non-null   float64
 6   Smoking                    2149 non-null   int64  
 7   AlcoholConsumption         2149 non-null   float64
 8   PhysicalActivity           2149 non-null   float64
 9   DietQuality                2149 non-null   float64
 10  SleepQuality               2149 non-null   float64
 11  FamilyHistoryAlzheimers    2149 non-null   int64  
 12  CardiovascularDisease      2149 non-null   int64  
 13  Diabetes                   2149 non-null   int64  
 14  Dep

## 2.2 Descriptive Statistics — Numeric Features

In [3]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
PatientID,2149.0,5825.000000,620.507185,4751.000000,5288.000000,5825.000000,6362.000000,6899.000000
Age,2149.0,74.908795,8.990221,60.000000,67.000000,75.000000,83.000000,90.000000
Gender,2149.0,0.506282,0.500077,0.000000,0.000000,1.000000,1.000000,1.000000
Ethnicity,2149.0,0.697534,0.996128,0.000000,0.000000,0.000000,1.000000,3.000000
EducationLevel,2149.0,1.286645,0.904527,0.000000,1.000000,1.000000,2.000000,3.000000
BMI,2149.0,27.655697,7.217438,15.008851,21.611408,27.823924,33.869778,39.992767
Smoking,2149.0,0.288506,0.453173,0.000000,0.000000,0.000000,1.000000,1.000000
AlcoholConsumption,2149.0,10.039442,5.757910,0.002003,5.139810,9.934412,15.157931,19.989293
PhysicalActivity,2149.0,4.920202,2.857191,0.003616,2.570626,4.766424,7.427899,9.987429
DietQuality,2149.0,4.993138,2.909055,0.009385,2.458455,5.076087,7.558625,9.998346


## 2.3 Descriptive Statistics — Categorical / Binary Features

In [4]:
binary_and_cat = [c for c in df.columns if df[c].nunique() <= 6 and c not in ['Diagnosis']]
for col in binary_and_cat:
    print(f"\n--- {col} ---")
    print(df[col].value_counts().sort_index())


--- Gender ---
Gender
0    1061
1    1088
Name: count, dtype: int64

--- Ethnicity ---
Ethnicity
0    1278
1     454
2     206
3     211
Name: count, dtype: int64

--- EducationLevel ---
EducationLevel
0    446
1    854
2    636
3    213
Name: count, dtype: int64

--- Smoking ---
Smoking
0    1529
1     620
Name: count, dtype: int64

--- FamilyHistoryAlzheimers ---
FamilyHistoryAlzheimers
0    1607
1     542
Name: count, dtype: int64

--- CardiovascularDisease ---
CardiovascularDisease
0    1839
1     310
Name: count, dtype: int64

--- Diabetes ---
Diabetes
0    1825
1     324
Name: count, dtype: int64

--- Depression ---
Depression
0    1718
1     431
Name: count, dtype: int64

--- HeadInjury ---
HeadInjury
0    1950
1     199
Name: count, dtype: int64

--- Hypertension ---
Hypertension
0    1829
1     320
Name: count, dtype: int64

--- MemoryComplaints ---
MemoryComplaints
0    1702
1     447
Name: count, dtype: int64

--- BehavioralProblems ---
BehavioralProblems
0    1812
1     33

## 2.4 Target Variable

`Diagnosis` is the label we're predicting: 1 = Alzheimer's diagnosed, 0 = not diagnosed.

In [5]:
target_counts = df[TARGET_COL].value_counts()
target_pct = df[TARGET_COL].value_counts(normalize=True) * 100
print(target_counts)
print()
print(target_pct.round(2))
print(f"\nClass imbalance ratio (majority:minority) ≈ {target_counts.max()/target_counts.min():.2f} : 1")

Diagnosis
0    1389
1     760
Name: count, dtype: int64

Diagnosis
0    64.63
1    35.37
Name: proportion, dtype: float64

Class imbalance ratio (majority:minority) ≈ 1.83 : 1


**Observation:** The dataset is moderately imbalanced (~65% No Diagnosis vs ~35% Diagnosis). This is not severe enough to require synthetic resampling (e.g. SMOTE), but it does mean:
- Accuracy alone would be a misleading metric.
- Model comparison in `06_Model_Development.ipynb` uses `class_weight='balanced'` where supported, and reports precision/recall/F1/ROC-AUC rather than accuracy alone.
- Stratified train/val/test splitting is essential (implemented in `src/preprocessing.py::split_data`) so every split preserves this ratio.

## 2.5 Identifier Columns (to be dropped before modelling)

In [6]:
print(df['PatientID'].nunique(), 'unique PatientIDs out of', len(df), 'rows')
print(df['DoctorInCharge'].unique())

2149 unique PatientIDs out of 2149 rows
<StringArray>
['XXXConfid']
Length: 1, dtype: str


`PatientID` is a unique row identifier and `DoctorInCharge` is a constant placeholder (`XXXConfid`, a redacted/confidentiality field) — both carry zero predictive signal and are dropped in `src/preprocessing.py::drop_non_predictive_columns`.

## 2.6 Value Range Sanity Check (clinical plausibility)

In [7]:
clinical_ranges = {
    'Age': (60, 90), 'MMSE': (0, 30), 'FunctionalAssessment': (0, 10), 'ADL': (0, 10),
    'SystolicBP': (90, 180), 'DiastolicBP': (60, 120), 'BMI': (15, 45),
}
for col, (lo, hi) in clinical_ranges.items():
    actual_lo, actual_hi = df[col].min(), df[col].max()
    status = "OK" if actual_lo >= lo and actual_hi <= hi else "CHECK"
    print(f"{col:25s} expected [{lo}, {hi}]  actual [{actual_lo:.1f}, {actual_hi:.1f}]  -> {status}")

Age                       expected [60, 90]  actual [60.0, 90.0]  -> OK
MMSE                      expected [0, 30]  actual [0.0, 30.0]  -> OK
FunctionalAssessment      expected [0, 10]  actual [0.0, 10.0]  -> OK
ADL                       expected [0, 10]  actual [0.0, 10.0]  -> OK
SystolicBP                expected [90, 180]  actual [90.0, 179.0]  -> OK
DiastolicBP               expected [60, 120]  actual [60.0, 119.0]  -> OK
BMI                       expected [15, 45]  actual [15.0, 40.0]  -> OK


**Summary of Data Understanding:**
- 2,149 patient records, 35 columns, no obvious structural issues at first pass.
- All values fall within clinically plausible ranges.
- Target is moderately imbalanced (~65/35) — informs modelling strategy.
- Two columns (`PatientID`, `DoctorInCharge`) will be dropped as non-predictive.

Formal missing-value, duplicate, and outlier profiling is done next in `03_Data_Profiling.ipynb`.